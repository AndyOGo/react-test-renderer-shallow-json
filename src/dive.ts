import type { ReactTestRendererTreeNodeFixed } from './types';

export function dive(
  tree: ReactTestRendererTreeNodeFixed,
  depth: number
): number {
  return typeof tree === 'string' || tree.nodeType === 'host'
    ? depth
    : depth - 1;
}
