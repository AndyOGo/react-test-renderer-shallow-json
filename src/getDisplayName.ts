import {
  ReactTestRendererTreeFixed,
  isContext,
  isNamedExoticComponentType,
} from './types';

const elementSymbolMap: Readonly<Record<symbol, string>> = {
  [Symbol.for('react.memo')]: 'Memo',
  [Symbol.for('react.context')]: 'Consumer',
  [Symbol.for('react.provider')]: 'Provider',
};

function appendSymbol(displayName: string, $$typeof?: symbol): string {
  if (!$$typeof || !elementSymbolMap[$$typeof]) {
    return displayName;
  }

  return `${displayName}.${elementSymbolMap[$$typeof]}`;
}

export function getDisplayName(
  type: ReactTestRendererTreeFixed['type'],
  fallback = '???'
): string {
  if (typeof type === 'string') {
    return type;
  }

  if (isContext(type)) {
    return appendSymbol(type._context.displayName || fallback, type.$$typeof);
  }

  if (isNamedExoticComponentType(type)) {
    return (
      type.displayName ||
      appendSymbol(
        type.type ? getDisplayName(type.type) : fallback,
        type.$$typeof
      )
    );
  }

  return type.displayName || type.name || fallback;
}
