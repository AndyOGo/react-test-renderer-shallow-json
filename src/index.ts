console.log('Start coding here...');
