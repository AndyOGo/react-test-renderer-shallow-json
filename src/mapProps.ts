import type { ReactTestRendererTreeFixed } from './types';

export function mapProps(props: ReactTestRendererTreeFixed['props']) {
  const { children, ...propsWithoutChildren } = props;

  return propsWithoutChildren;
}
