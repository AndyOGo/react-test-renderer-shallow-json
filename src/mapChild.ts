import type {
  ReactTestRendererNodeFixed,
  ReactTestRendererTreeChild,
} from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';

export function mapChild(
  child: ReactTestRendererTreeChild
): ReactTestRendererNodeFixed {
  if (typeof child === 'string') {
    return child;
  }

  const { key, ref } = child;

  const json = {
    type: getDisplayName(child.type),
    props: mapProps({
      ...child.props,
      ...(key && { key }),
      ...(ref && { ref }),
    }),
    children: mapChildren(child.props),
  };

  Object.defineProperty(json, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return json;
}
