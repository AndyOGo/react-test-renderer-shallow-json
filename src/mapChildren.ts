import type {
  ReactTestRendererJSONFixed,
  ReactTestRendererTreeFixed,
} from './types';
import { mapChild } from './mapChild';

export function mapChildren(
  props: ReactTestRendererTreeFixed['props']
): ReactTestRendererJSONFixed['children'] {
  const { children } = props;

  if (!children) {
    return null;
  }

  if (!Array.isArray(children)) {
    return [mapChild(children)];
  }

  return children.map(mapChild);
}
