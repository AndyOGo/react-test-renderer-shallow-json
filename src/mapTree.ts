import type { ReactTestRendererNode } from 'react-test-renderer';
import type { ReactTestRendererTreeNodeFixed } from './types';
import { getDisplayName } from './getDisplayName';
import { mapProps } from './mapProps';
import { mapChildren } from './mapChildren';
import { mapTrees } from './mapTrees';
import { withReactTestSymbol } from './withReactTestSymbol';

export function mapTree(
  tree: ReactTestRendererTreeNodeFixed,
  depth = 1
): ReactTestRendererNode {
  if (typeof tree === 'string') {
    return tree;
  }

  const json = {
    type: getDisplayName(tree.type),
    props: mapProps(tree.props),
    children:
      depth > 0 ? mapTrees(tree.rendered, depth) : mapChildren(tree.props),
  };

  return withReactTestSymbol(json) as ReactTestRendererNode;
}
