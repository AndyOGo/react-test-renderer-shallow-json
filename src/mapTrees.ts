import type { ReactTestRendererJSON } from 'react-test-renderer';
import type { ReactTestRendererTreeFixed } from './types';
import { mapTree } from './mapTree';
import { dive } from './dive';

export function mapTrees(
  trees: ReactTestRendererTreeFixed['rendered'],
  depth = 1
): ReactTestRendererJSON['children'] {
  if (!trees) {
    return null;
  }

  if (Array.isArray(trees)) {
    return trees.map((tree) => mapTree(tree, dive(tree, depth)));
  }

  return [mapTree(trees, dive(trees, depth))];
}
