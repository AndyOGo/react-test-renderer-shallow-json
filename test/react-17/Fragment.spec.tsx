import React, { Fragment } from 'react';
import { create } from 'react-test-renderer';
import { shallowJSON } from '../../src';
import { TestWrapper } from './TestWrapper';

describe('Fragment', () => {
  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <>
            <div />
            <TestWrapper />
          </>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot', () => {
    expect(
      shallowJSON(
        create(
          <Fragment>
            <div />
            <TestWrapper />
          </Fragment>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <>
              <div />
              <TestWrapper />
            </>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot as children', () => {
    expect(
      shallowJSON(
        create(
          <TestWrapper>
            <Fragment>
              <div />
              <TestWrapper />
            </Fragment>
          </TestWrapper>
        ).toTree()
      )
    ).toMatchSnapshot();
  });
});
