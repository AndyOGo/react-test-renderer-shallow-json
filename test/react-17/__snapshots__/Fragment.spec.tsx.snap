// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragment matches snapshot 1`] = `
[
  <div />,
  <TestWrapper>
    <div />
  </TestWrapper>,
]
`;

exports[`Fragment matches snapshot 2`] = `
[
  <div />,
  <TestWrapper>
    <div />
  </TestWrapper>,
]
`;

exports[`Fragment matches snapshot as children 1`] = `
<TestWrapper>
  <div>
    <React.Fragment>
      <div />
      <TestWrapper />
    </React.Fragment>
  </div>
</TestWrapper>
`;

exports[`Fragment matches snapshot as children 2`] = `
<TestWrapper>
  <div>
    <React.Fragment>
      <div />
      <TestWrapper />
    </React.Fragment>
  </div>
</TestWrapper>
`;
